### PR TITLE
Select Smarter Lags for Time Series

### LINKED ISSUE #2733: Smarter lagging for time series
Right now, the `DelayedFeaturesTransformer` computes every lag from 1 to max_delay. Not every lag is meaningful and this also greatly increases the size of the data for large max_delay values. I think a more intelligent (and data-efficient approach) would be to use the date frequency to infer useful lag frequencies:

For example, if the data is daily use a mix of daily and weekly lags. If the data is monthly, use monthly and yearly lags.

This is a complex feature and there is likely not a good "one-size-fits-all" lagging strategy. This is just a half-way thought I had that I wanted to document. This fits within #2510 , so it's possible we decide to go in a different direction in the future.

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description
Fixes #2733 

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: evalml/tests/component_tests/test_delayed_features_transformer.py [hunk 1]
```diff
@@ -84,12 +88,11 @@ def test_delayed_feature_extractor_maxdelay3_forecasthorizon1_gap0(
...
         }
     )
-
     assert_frame_equal(
         answer,
-        DelayedFeatureTransformer(max_delay=3, gap=0, forecast_horizon=1).fit_transform(
-            X=X, y=y
-        ),
+        DelayedFeatureTransformer(
+            max_delay=3, gap=0, forecast_horizon=1, conf_level=1.0
```
Comment (freddyaboulton) [inline evalml/tests/component_tests/test_delayed_features_transformer.py:94]: Using a `conf_level` of 1.0 selects all lags. Some of these tests rely on the DelayedFeatureTransformer selecting all lags.

**Review by bchen1116 — CHANGES_REQUESTED** [review]
Great work on this! You got this out so fast. I think one thing that we should add before merging this would be some documentation, maybe [here](https://evalml.alteryx.com/en/stable/user_guide/timeseries.html), that would allow users to understand what this component is doing and how changing conf_level could change the lags used. This would help with un-black-boxing the component.

Otherwise, left some nits and some questions. Good stuff though!

File: evalml/pipelines/components/transformers/preprocessing/delayed_feature_transformer.py [hunk 2]
```diff
@@ -14,6 +18,9 @@ class DelayedFeatureTransformer(Transformer):
         date_index (str): Name of the column containing the datetime information used to order the data. Ignored.
         max_delay (int): Maximum number of time units to delay each feature. Defaults to 2.
         forecast_horizon (int): The number of time periods the pipeline is expected to forecast.
+        conf_level (float, None): Float between 0 and 1 that determines the confidence interval size used to select
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/preprocessing/delayed_feature_transformer.py:21]: Absolutely. Will get rid of the outdated `None` as well as adding some validation that conf_level is in the range (0, 1].

File: evalml/pipelines/components/transformers/preprocessing/delayed_feature_transformer.py [hunk 3]
```diff
@@ -91,6 +105,29 @@ def _encode_X_while_preserving_index(X_categorical):
...
+            # Compute the acf and find its peaks
+            acf_values, ci_intervals = acf(
+                y, nlags=len(y) - 1, fft=True, alpha=conf_level
+            )
+            peaks, _ = find_peaks(acf_values)
+
+            # Return lags that are significant peaks or the first 10 significant lags
+            index = np.arange(len(acf_values))
+            significant = np.logical_or(ci_intervals[:, 0] > 0, ci_intervals[:, 1] < 0)
+            first_significant_10 = index[:10][significant[:10]]
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/preprocessing/delayed_feature_transformer.py:121]: Yea I can see why this is confusing now. The intention is to select the significant lags in the range [0, 10] not the first 10 significant lags.

File: evalml/pipelines/components/transformers/preprocessing/delayed_feature_transformer.py [hunk 4]
```diff
@@ -14,6 +18,9 @@ class DelayedFeatureTransformer(Transformer):
         date_index (str): Name of the column containing the datetime information used to order the data. Ignored.
         max_delay (int): Maximum number of time units to delay each feature. Defaults to 2.
         forecast_horizon (int): The number of time periods the pipeline is expected to forecast.
+        conf_level (float, None): Float between 0 and 1 that determines the confidence interval size used to select
+            which lags to compute from the set of [1, max_delay]. A delay of 1 will always be computed. If 1,
+            selects all possible lags in the set of [1, max_delay], inclusive.
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/preprocessing/delayed_feature_transformer.py:23]: Good catch. Added some parameter validation as well.

File: evalml/tests/component_tests/test_delayed_features_transformer.py [hunk 5]
```diff
@@ -413,22 +419,125 @@ def test_delay_feature_transformer_y_is_none(delayed_features_data):
...
+    # Although the conf_level is hard-coded we mock the return values of
+    # find_peaks and acf so that we simulate different significant lags.
+    dft = DelayedFeatureTransformer(
+        max_delay=MAX_DELAY, forecast_horizon=FORECAST_HORIZON, conf_level=0.05, gap=0
+    )
+    new_X = dft.fit_transform(X, y)
+
+    first_significant_10 = [l for l in significant_lags if l < 10]
+    expected_lags = (
+        set(significant_lags + peaks).intersection(peaks).union(first_significant_10)
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/tests/component_tests/test_delayed_features_transformer.py:479]: Yep, making this change now!

**Comment by freddyaboulton** [comment]
@bchen1116 Added details on the algorithm to the component docstring and added a link from the user guide to the component reference since I didn't want to jump deep into the specifics of a component in the user guide. Thanks for the feedback though, the documentation definitely needed improvement.

**Review by ParthivNaresh — APPROVED** [review]
Looks good! In your perf tests it showed that `Daily Female Births` had slightly lower validation results while `Southern Oscillations` had slightly better. It would be interesting to see how Prophet and ARIMA perform (once we enable DelayedFeatureTransformer for ARIMA).

File: evalml/pipelines/components/transformers/preprocessing/delayed_feature_transformer.py [hunk 6]
```diff
@@ -91,6 +126,32 @@ def _encode_X_while_preserving_index(X_categorical):
             index=X_categorical.index,
         )
 
+    @staticmethod
+    def _find_significant_lags(y, conf_level, max_delay):
+        all_lags = np.arange(max_delay + 1)
+        if conf_level is not None and y is not None:
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/preprocessing/delayed_feature_transformer.py:132]: I think what happened originally I set the default value to `None` to mean "no lags" but that doesn't work well with our tuners so I made it be a float. I'll move the none check to the init! Thank you.

**Review by bchen1116 — APPROVED** [review]
LGTM! Thanks for making the changes! Agreed with @ParthivNaresh about `conf_level=None`, but other than that, great work!

**Pull request merged**